ocaml/xapi/xapi_pif.ml

```diff
@@ -767,6 +767,31 @@ let set_property ~__context ~self ~name ~value =
         Nm.bring_pif_up ~__context pif
     ) (self :: vlan_pifs)
 
+let pif_has_clusters ~__context (self : API.ref_PIF) =
+  debug "Looking up clusters for PIF: %s" (Ref.string_of self);
+  let network_of_pif =
+    try Db.PIF.get_network ~__context ~self with _ ->
+      failwith (Printf.sprintf "Could not find network of PIF: %s" (Ref.string_of self))
+  in
+  info "Got network: %s" (Ref.string_of network_of_pif);
+  (Db.Cluster.get_records_where ~__context
+    ~expr:Db_filter_types.(Eq(Literal (Ref.string_of network_of_pif),Field "network")))
+    |> function
+    | []       -> false
+    | (a,_)::_ -> true
+
+let assert_no_clustering_enabled ~__context ~self =
+  if pif_has_clusters ~__context self
+  then raise Api_errors.(Server_error(Api_errors.cannot_change_pif_properties, [Ref.string_of self]))
```

Comment:
One issue with this exception is that its description currently talks about bonded PIFs (in `datamodel.ml`)

Reply:
Let's introduce a new API error that is specific to clustering.

Reply:
Minor: you can remove the second mention of `Api_errors.`.